src/learner.cc

```diff
@@ -601,7 +615,7 @@ class LearnerImpl : public Learner {
       num_feature = std::max(num_feature, static_cast<unsigned>(num_col));
     }
     // run allreduce on num_feature to find the maximum value
-    rabit::Allreduce<rabit::op::Max>(&num_feature, 1);
+    rabit::Allreduce<rabit::op::Max>(&num_feature, 1, nullptr, nullptr, true);
```

Comment:
That;s really hacky ....

Reply:
This is a trade off we made to allow adhoc/one time allreduce sync. We might optimize this out after we unify "adhoc/one-time" operations with "inloop" operations with help from caller signature. But this is not a trivial task to decide which operation result should be evict verses which should be kept. One thing at a time.
 
detail design here
https://github.com/dmlc/xgboost/files/3482936/Rabit.bootstrap.operations.support.pdf

pr
https://github.com/dmlc/rabit/pull/98

Reply:
update, I manage to get rid of this, after merge to master api will be same as before and user don't need to worry about cache or non cache
```
rabit::Allreduce<rabit::op::Max>(&num_feature, 1);
```
https://github.com/dmlc/rabit/pull/102/commits/69789ec4510d019d25127c62973acef10d10c6fc